### PR TITLE
fix: Allow whitespace to delimit matrix specifications, better support parsing  identifiers followed by a minus operator.

### LINKED ISSUE #284: Identifiers minus expressions fail to parse without whitespace
[pyQuil#1658](https://github.com/rigetti/pyquil/issues/1658) reported a couple of issues with parsing. One of them occurs when parsing expressions in this DEFGATE:

```
DEFGATE CAN(%alpha, %beta, %gamma):
    (cis((%alpha+%beta-%gamma)/2)+cis((%alpha-%beta+%gamma)/2))/2, 0, 0, (cis((%alpha-%beta+%gamma)/2)-cis((%alpha+%beta-%gamma)/2))/2
    0, (cis((%alpha+%beta+%gamma)/(-2))+cis((%beta+%gamma-%alpha)/2))/2, (cis((%alpha+%beta+%gamma)/(-2))-cis((%beta+%gamma-%alpha)/2))/2, 0
    0, (cis((%alpha+%beta+%gamma)/(-2))-cis((%beta+%gamma-%alpha)/2))/2, (cis((%alpha+%beta+%gamma)/(-2))+cis((%beta+%gamma-%alpha)/2))/2, 0
    (cis((%alpha-%beta+%gamma)/2)-cis((%alpha+%beta-%gamma)/2))/2, 0, 0, (cis((%alpha+%beta-%gamma)/2)+cis((%alpha-%beta+%gamma)/2))/2
```

In particular, the expression `%beta-%gamma` fails to pass the lexer.

In Quil `-` is the minus operator and is also valid to use in an identifier, as long as it isn't the first or last character in the identifier. In this case, it should be lexed as a minus operator, but it is instead lexes it as part of the `beta` identifier. Since that results in an invalid identifier "beta-", the lexer returns an error.

### PR DESCRIPTION
**Pull request by MarquessV**
closes #284 and #285

### CODE REVIEW

File: quil-rs/src/parser/common.rs [hunk 1]
```diff
@@ -178,7 +178,10 @@ pub(crate) fn parse_matrix<'a>(
             token!(NewLine),
             preceded(
                 token!(Indentation),
-                separated_list0(token!(Comma), parse_expression),
+                separated_list0(
+                    pair(token!(Comma), many0(token!(Indentation))),
```
Comment (MarquessV) [inline quil-rs/src/parser/common.rs:182]: This fix might be a little too targeted, but changing how indentation is handled is a much more heavy handed fix.

The Quil spec states that indentation is exactly four spaces and can only occur at the start of a line. The new line is part of the definition of indentation, which means that indenting more than once in a single line isn't possible. It's unclear if this is intentional or not. Nothing in the Quil spec states that nested definitions are invalid, but they would require multiple levels of indentation.

File: quil-rs/src/parser/lexer/mod.rs [hunk 2]
```diff
@@ -262,21 +262,17 @@ fn lex_identifier_raw(input: LexInput) -> InternalLexResult<String> {
...
+                take_while(is_valid_identifier_end_character),
+                recognize(many0(pair(
+                    take_while1(is_dash),
+                    take_while1(is_valid_identifier_end_character),
+                ))),
             )),
-            |(left, right)| format!("{left}{right}"),
+            |(leading, middle, trailing_dash_vars)| {
+                format!("{leading}{middle}{trailing_dash_vars}")
+            },
```
Comment (MarquessV) [inline quil-rs/src/parser/lexer/mod.rs:273]: The idea here is to parse any valid identifier up to any dashes. After that point, any dash should be followed by at least one valid end character for an identifier. Any such pairs are returned and appended to the resulting identifier.

File: quil-rs/src/parser/common.rs (hunk 1 above)
Comment (jselig-rigetti) [reply quil-rs/src/parser/common.rs:182]: Is this only when someone wants to align the non-leftmost items within a matrix? Would it necessarily be a `Indentation` token or could it be any amount of whitespace?

File: quil-rs/src/parser/lexer/mod.rs (hunk 2 above)
Comment (jselig-rigetti) [reply quil-rs/src/parser/lexer/mod.rs:273]: > After that point, any dash should be followed by at least one valid end character for an identifier

"any dash" or "any dash sequence"? If not the latter, then https://github.com/rigetti/quil-rs/pull/286/files#diff-a5bcc7a39fd5fe75b07e825fc96eeaa32b6cbed091f5936b09d97f9e9112769eR267 seems off

Comment (MarquessV) [reply quil-rs/src/parser/lexer/mod.rs:273]: Yeah, the latter, a singular dash or a sequence of dashes are both valid, as long a valid end character comes after.

File: quil-rs/src/parser/common.rs (hunk 1 above)
Comment (MarquessV) [reply quil-rs/src/parser/common.rs:182]: This use case is just for aligning items in the matrix. The lexer conisders a tab character or four consecutive spaces as a single indentation token. Other whitespace is ignored.

**Review by jselig-rigetti — APPROVED** [review]

File: quil-rs/src/parser/common.rs (hunk 1 above)
Comment (jselig-rigetti) [reply quil-rs/src/parser/common.rs:182]: oops - I'll unresolve since this was your thread originally

**Pull request merged**